sdetools/extlib/sqlitedict.py

```
@@ -0,0 +1,520 @@
+        # `select count (*)` is super slow in sqlite (does a linear scan!!)
+        # As a result, len() is very slow too once the table size grows beyond trivial.
+        # We could keep the total count of rows ourselves, by means of triggers,
+        # but that seems too complicated and would slow down normal operation
+        # (insert/delete etc).
+        GET_LEN = 'SELECT COUNT(*) FROM "%s"' % self.tablename
+        rows = self.conn.select_one(GET_LEN)[0]
+        return rows if rows is not None else 0
+
+    def __bool__(self):
+        # No elements is False, otherwise True
+        GET_MAX = 'SELECT MAX(ROWID) FROM "%s"' % self.tablename
+        m = self.conn.select_one(GET_MAX)[0]
+        # Explicit better than implicit and bla bla
+        return True if m is not None else False
+
+    def iterkeys(self):
+        GET_KEYS = 'SELECT key FROM "%s" ORDER BY rowid' % self.tablename
+        for key in self.conn.select(GET_KEYS):
+            yield key[0]
+
+    def itervalues(self):
+        GET_VALUES = 'SELECT value FROM "%s" ORDER BY rowid' % self.tablename
+        for value in self.conn.select(GET_VALUES):
+            yield decode(value[0])
+
+    def iteritems(self):
+        GET_ITEMS = 'SELECT key, value FROM "%s" ORDER BY rowid' % self.tablename
+        for key, value in self.conn.select(GET_ITEMS):
+            yield key, decode(value)
+
+    def keys(self):
+        return self.iterkeys() if major_version > 2 else list(self.iterkeys())
+
+    def values(self):
+        return self.itervalues() if major_version > 2 else list(self.itervalues())
+
+    def items(self):
+        return self.iteritems() if major_version > 2 else list(self.iteritems())
+
+    def __contains__(self, key):
+        HAS_ITEM = 'SELECT 1 FROM "%s" WHERE key = ?' % self.tablename
+        return self.conn.select_one(HAS_ITEM, (key,)) is not None
+
+    def __getitem__(self, key):
+        GET_ITEM = 'SELECT value FROM "%s" WHERE key = ?' % self.tablename
```

Do we want to limit how many items we get back. We're only using the first one anyway. Do we want to raise an error if we get more than one?